File: win/system/backend/app/api/members.py
"""
Member API endpoints
Handles member listing, details, and tags
"""

from typing import List
from fastapi import APIRouter, HTTPException, Depends, Body

from app.services.pluralkit import get_members
from app.services.tag_service import (
    get_member_tags, update_member_tags, add_member_tag,
    remove_member_tag, enrich_members_with_tags
)
from app.services.status_service import enrich_members_with_status
from app.dependencies.auth import get_admin_user
from app.utils.cache import set_in_cache

router = APIRouter()

@router.get("/members")
async def list_members():
    """Get all members with tags and status information"""
    try:
        # Get members from PluralKit
        members_data = await get_members()
        
        # Enrich with tags
        members_with_tags = enrich_members_with_tags(members_data)
        
        # Enrich with status information
        members_with_status = enrich_members_with_status(members_with_tags)
        
        return members_with_status
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to fetch members: {str(e)}"
        )

@router.get("/member/{member_id}")
async def member_detail(member_id: str):
    """Get details for a specific member"""
    try:
        members = await get_members()
        
        # Find member by ID or name
        member = None
        for m in members:
            if m["id"] == member_id or m["name"].lower() == member_id.lower():
                member = m
                break
        
        if not member:
            raise HTTPException(
                status_code=404,
                detail="Member not found"
            )
        
        # Enrich with tags and status
        member_with_tags = enrich_members_with_tags([member])[0]
        member_with_status = enrich_members_with_status([member_with_tags])[0]
        
        return member_with_status
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to fetch member details: {str(e)}"
        )

# ============================================================================
# MEMBER TAGS ENDPOINTS
# ============================================================================

@router.get("/member-tags")
async def list_member_tags(user=Depends(get_admin_user)):
    """Get all member tag assignments (admin only)"""
    try:
        member_tags = get_member_tags()
        return {
            "status": "success",
            "member_tags": member_tags
        }
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to fetch member tags: {str(e)}"
        )

@router.post("/member-tags/{member_identifier}")
async def update_member_tag_list(
    member_identifier: str,
    tags: List[str] = Body(...),
    user=Depends(get_admin_user)
):
    """Update the complete tag list for a member (admin only)"""
    try:
        success = update_member_tags(member_identifier, tags)
        
        if success:
            # Clear member cache to reflect changes
            set_in_cache("members_raw", None, 0)
            
            return {
                "status": "success",
                "message": f"Updated tags for {member_identifier}",
                "tags": tags
            }
        else:
            raise HTTPException(
                status_code=500,
                detail="Failed to update member tags"
            )
            
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to update member tags: {str(e)}"
        )

@router.post("/member-tags/{member_identifier}/add")
async def add_single_member_tag(
    member_identifier: str,
    tag: str = Body(..., embed=True),
    user=Depends(get_admin_user)
):
    """Add a single tag to a member (admin only)"""
    try:
        success = add_member_tag(member_identifier, tag)
        
        if success:
            # Clear member cache
            set_in_cache("members_raw", None, 0)
            
            return {
                "status": "success",
                "message": f"Added tag '{tag}' to {member_identifier}"
            }
        else:
            return {
                "status": "info",
                "message": f"Tag '{tag}' already exists for {member_identifier}"
            }
            
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to add member tag: {str(e)}"
        )

@router.delete("/member-tags/{member_identifier}/{tag}")
async def remove_single_member_tag(
    member_identifier: str,
    tag: str,
    user=Depends(get_admin_user)
):
    """Remove a single tag from a member (admin only)"""
    try:
        success = remove_member_tag(member_identifier, tag)
        
        if success:
            # Clear member cache
            set_in_cache("members_raw", None, 0)
            
            return {
                "status": "success",
                "message": f"Removed tag '{tag}' from {member_identifier}"
            }
        else:
            raise HTTPException(
                status_code=404,
                detail=f"Tag '{tag}' not found for {member_identifier}"
            )
            
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to remove member tag: {str(e)}"
        )

File: co/system/backend/src/routes/members.ts
/**
 * Copyright (c) 2026 Clove Twilight
 * Licensed under the ESAL-2.0 Licence.
 * See LICENCE.md in the project root for full licence information.
 */

import { Router, type Request, type Response, type NextFunction } from 'express';
import { z } from 'zod';

import { getMembers } from '../services/pluralkit.js';
import {
  getMemberTags,
  updateMemberTags,
  addMemberTag,
  removeMemberTag,
  enrichMembersWithTags,
} from '../services/tag_service.js';
import { enrichMembersWithStatus } from '../services/status_service.js';
import { requireAuth, requireAdmin } from '../dependencies/auth.js';
import { setInCache } from '../utils/cache.js';
import { HttpError } from '../core/errors.js';

export const membersRouter = Router();

/** Get all members with tags and status information */
membersRouter.get('/members', async (_req: Request, res: Response, next: NextFunction) => {
  try {
    // Get members from PluralKit
    const membersData = await getMembers();

    // Enrich with tags
    const membersWithTags = await enrichMembersWithTags(membersData);

    // Enrich with status information
    const membersWithStatus = await enrichMembersWithStatus(membersWithTags);

    res.json(membersWithStatus);
  } catch (err) {
    next(new HttpError(500, `Failed to fetch members: ${String(err)}`));
  }
});

/** Get details for a specific member */
membersRouter.get('/member/:member_id', async (req: Request, res: Response, next: NextFunction) => {
  try {
    const memberId = req.params.member_id;
    const members = await getMembers();

    // Find member by ID or name
    const member = members.find(
      (m) => m.id === memberId || String(m.name ?? '').toLowerCase() === memberId.toLowerCase(),
    );

    if (!member) {
      next(new HttpError(404, 'Member not found'));
      return;
    }

    // Enrich with tags and status
    const [memberWithTags] = await enrichMembersWithTags([member]);
    const [memberWithStatus] = await enrichMembersWithStatus([memberWithTags]);

    res.json(memberWithStatus);
  } catch (err) {
    if (err instanceof HttpError) {
      next(err);
      return;
    }
    next(new HttpError(500, `Failed to fetch member details: ${String(err)}`));
  }
});

// ============================================================================
// MEMBER TAGS ENDPOINTS
// ============================================================================

/** Get all member tag assignments (admin only) */
membersRouter.get(
  '/member-tags',
  requireAuth,
  requireAdmin,
  async (_req: Request, res: Response, next: NextFunction) => {
    try {
      const memberTags = await getMemberTags();
      res.json({ status: 'success', member_tags: memberTags });
    } catch (err) {
      next(new HttpError(500, `Failed to fetch member tags: ${String(err)}`));
    }
  },
);

/** Update the complete tag list for a member (admin only) */
membersRouter.post(
  '/member-tags/:member_identifier',
  requireAuth,
  requireAdmin,
  async (req: Request, res: Response, next: NextFunction) => {
    const parsed = z.array(z.string()).safeParse(req.body);
    if (!parsed.success) {
      res.status(422).json({ detail: parsed.error.issues });
      return;
    }
    const tags = parsed.data;
    const memberIdentifier = req.params.member_identifier;

    try {
      const success = await updateMemberTags(memberIdentifier, tags);

      if (success) {
        // Clear member cache to reflect changes
        setInCache('members_raw', null, 0);

        res.json({ status: 'success', message: `Updated tags for ${memberIdentifier}`, tags });
      } else {
        next(new HttpError(500, 'Failed to update member tags'));
      }
    } catch (err) {
      next(new HttpError(500, `Failed to update member tags: ${String(err)}`));
    }
  },
);

/** Add a single tag to a member (admin only) */
membersRouter.post(
  '/member-tags/:member_identifier/add',
  requireAuth,
  requireAdmin,
  async (req: Request, res: Response, next: NextFunction) => {
    const tag = req.body?.tag;
    const memberIdentifier = req.params.member_identifier;

    if (typeof tag !== 'string') {
      res.status(422).json({ detail: "Body must include a 'tag' string field" });
      return;
    }

    try {
      const success = await addMemberTag(memberIdentifier, tag);

      if (success) {
        // Clear member cache
        setInCache('members_raw', null, 0);

        res.json({ status: 'success', message: `Added tag '${tag}' to ${memberIdentifier}` });
      } else {
        res.json({ status: 'info', message: `Tag '${tag}' already exists for ${memberIdentifier}` });
      }
    } catch (err) {
      next(new HttpError(500, `Failed to add member tag: ${String(err)}`));
    }
  },
);

/** Remove a single tag from a member (admin only) */
membersRouter.delete(
  '/member-tags/:member_identifier/:tag',
  requireAuth,
  requireAdmin,
  async (req: Request, res: Response, next: NextFunction) => {
    const { member_identifier: memberIdentifier, tag } = req.params;

    try {
      const success = await removeMemberTag(memberIdentifier, tag);

      if (success) {
        // Clear member cache
        setInCache('members_raw', null, 0);

        res.json({ status: 'success', message: `Removed tag '${tag}' from ${memberIdentifier}` });
      } else {
        next(new HttpError(404, `Tag '${tag}' not found for ${memberIdentifier}`));
      }
    } catch (err) {
      if (err instanceof HttpError) {
        next(err);
        return;
      }
      next(new HttpError(500, `Failed to remove member tag: ${String(err)}`));
    }
  },
);